### PR TITLE
port online_offline tests to work on Linux 

### PR DESCRIPTION
**Pull request by jsalinasintel**
Added two routines to be used by online_offline and included fixes
for Linux paltform

### CODE REVIEW

File: tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh [hunk 1]
```diff
@@ -60,13 +61,13 @@ function cleanup
...
 
-	$KILL $killpid >/dev/null 2>&1
+	kill_process $killpid
 	[[ -e $TESTDIR ]] && log_must $RM -rf $TESTDIR/*
 }
 
 log_assert "Turning a disk offline and back online during I/O completes."
 
-$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+start_io 67108864 8192 $TESTDIR/$TESTFILE1
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh]: Please leave this as $((64 * 1024 * 1024)) which is clearer.

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh [hunk 2]
```diff
@@ -61,15 +62,15 @@ function cleanup
...
+	kill_process $killpid
 	[[ -e $TESTDIR ]] && log_must $RM -rf $TESTDIR/*
 }
 
 log_assert "Turning both disks offline should fail."
 
 log_onexit cleanup
 
-$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+start_io 67108864 8192 $TESTDIR/$TESTFILE1
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: Same

File: tests/zfs-tests/include/libtest.shlib [hunk 3]
```diff
@@ -37,7 +37,7 @@
 
 function is_linux
 {
-	if [[ $($UNAME -o) == "GNU/Linux" ]]; then
+	if [[ $($UNAME -s) == "Linux" ]]; then
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: Let's split this unrelated change in to its own commit.

File: tests/zfs-tests/include/libtest.shlib [hunk 4]
```diff
@@ -2843,3 +2843,30 @@ function block_device_wait
...
+{
+	typeset -l i killpid=$1
+
+	#attempt to kill
+	if $PS --no-headers -p $killpid > /dev/null; then
+		log_must $KILL $killpid
+	fi
+	#verify killed worked
+	if $PS --no-headers -p $killpid > /dev/null; then
+		log_mustnot $KILL $killpid
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: This isn't a safe way to check.  It's entirely possible that after the original process died it's pid was recycled for a new unrelated process.

File: tests/zfs-tests/include/libtest.shlib [hunk 5]
```diff
@@ -2843,3 +2843,30 @@ function block_device_wait
...
+{
+	typeset -l i sizetouse=$1
+	typeset -l i blocksize=$2
+	typeset -l i filename=$3
+
+	#start some IO
+	#This used to be: $FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+	dd_string="$DD if=/dev/urandom iflag=fullblock bs=$blocksize count=$sizetouse of=$filename & "
+	log_note "$dd_string"
+	dd_string
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: The fact that `start_io` will start a background process isn't at all clear from the function name.  If we're going to add helpers for this kind of thing let's make it clear this is async.  Call it say `start_io_async`.

Why exactly can't FILE_TRUNC be used for this?  It would be preferable to using `/dev/urandom` which is meant for small chunks of random data and will be exceptionally slow compared to the alternatives.

Why not use `log_must` for this dd for logging purposes?

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh [hunk 6]
```diff
@@ -30,6 +30,7 @@
 #
 
 . $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/online_offline/online_offline.cfg
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: It appears that almost all of the exports in the cfg file are unused.  There is no equivalent .cfg file in OpenZFS for this test files.  Adding a .cfg files makes sense when it's appropriate but I don' t think it should be mandatory.

File: tests/zfs-tests/include/libtest.shlib [hunk 7]
```diff
@@ -2843,3 +2843,32 @@ function block_device_wait
 		$UDEVADM settle
 	fi
 }
+
+function kill_process
```
Comment (rlaager) [inline tests/zfs-tests/include/libtest.shlib]: There is stray trailing whitespace on a number of lines. Give these things a: sed -i "s|[[:space:]]\+$||" FILE

File: tests/zfs-tests/include/libtest.shlib [hunk 8]
```diff
@@ -2844,29 +2844,31 @@ function block_device_wait
...
+function kill_process 
+{ 
 	typeset -l i killpid=$1
 
-	#attempt to kill
+	#attempt to kill 
+	log_note "$PS --no-headers -p $killpid"
 	if $PS --no-headers -p $killpid > /dev/null; then
-		log_must $KILL $killpid
+        	log_must $KILL $killpid
```
Comment (rlaager) [inline tests/zfs-tests/include/libtest.shlib]: This mixes tabs and spaces. This file is using tabs to indent.

File: tests/zfs-tests/include/libtest.shlib [hunk 9]
```diff
@@ -2844,29 +2844,31 @@ function block_device_wait
...
 	fi
-	#verify killed worked
+	#verify killed worked 
 	if $PS --no-headers -p $killpid > /dev/null; then
-		log_mustnot $KILL $killpid
+        	log_mustnot $KILL $killpid
```
Comment (rlaager) [inline tests/zfs-tests/include/libtest.shlib]: This mixes tabs and spaces. This file is using tabs to indent.

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh (hunk 6 above)
Comment (jsalinasintel) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: If the cfg file is not used it does not work for physical disks -- we are doing all of our testing with physical disks so we need to use the cfg file or make a default that works with physical disks

File: tests/zfs-tests/include/libtest.shlib (hunk 3 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: that is a great idea - will do

File: tests/zfs-tests/include/libtest.shlib (hunk 4 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: When we talked about this during the hackathon the idea that came was to check the name as well.  I suppose we could also increase pid_max.  Let me know what your thoughts are -- I am open either, both, or something new I haven't thought of

File: tests/zfs-tests/include/libtest.shlib (hunk 5 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: Yes, we should use better function names.  Yes, log_must makes since for starting a command. 

Do you think FILE_TRUNC is the best tool for an IO load?  We can make a dd command line that works on Linux, Darwin, and freebsd.  But we can't for FILE_TRUNC.  On Linux it returns too quickly we want an IO load, not something that returns quickly.   

We looked at other options like shred but they did not prove to be portable to other platforms.  My suggestion was to write some pattern IO and we could use dd or not.

File: tests/zfs-tests/include/libtest.shlib (hunk 8 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: Hmm I will look at this I must have unloaded the vim macros at some point

File: tests/zfs-tests/include/libtest.shlib (hunk 9 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: Okay thank you for pointing this out I will fix this.

**Comment by jsalinasintel** [comment]
I would probably not be inclined to add a new test to linux.run.  Rather I would get some other people to run it and if they are successful then add it in.  I don't want to force others to run software that hasn't been run in more than one place.

File: tests/zfs-tests/include/libtest.shlib (hunk 4 above)
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: I'd suggest use ps to look from the previous pid and if found checking the name.

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh (hunk 6 above)
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: I don't follow.  Why should it matter for a physical disk config if the default values appear in `online_offline.cfg` or the test case itself?

File: tests/zfs-tests/include/libtest.shlib (hunk 5 above)
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: The test suite itself provides a very simple tool called `file_write` which can be used to generate an IO load with a simple pattern.  And since it part of the test suite we can make it do whatever we want and then have that functionality on all platforms.

Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: file_write looks perfect for what we want to do -- I had never heard of this thank you for passing it along

File: tests/zfs-tests/include/libtest.shlib (hunk 4 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: I have this in the routine will push into PR once I have added file_write as well

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh (hunk 6 above)
Comment (jsalinasintel) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: Actually I may have misunderstood what you were saying.  We didn't create the config file it was already there.  We just made it be sourced because when when the config file is not source it only works for loop back devices.  We want this test to work for both physical disks and loopback devices.  But, yes we could purge these I believe: xport HOLES_FILESIZE=${HOLES_FILESIZE-"67108864"} # 64 Mb                      
export HOLES_BLKSIZE=${HOLES_BLKSIZE-"512"}                                     
export HOLES_SEED=${HOLES_SEED-""}                                              
export HOLES_FILEOFFSET=${HOLES_FILEOFFSET-""}                                  
export HOLES_COUNT=${HOLES_COUNT-"16384"}      # FILESIZE/BLKSIZE/8

**Comment by jsalinasintel** [comment]
Not sure what to do with this the required the test that failed looks like it is out of space and completely unrelated to this change.  Perhaps tests should re-trigger after space has been cleared?

File: tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh [hunk 10]
```diff
@@ -60,13 +61,21 @@ function cleanup
...
 log_assert "Turning a disk offline and back online during I/O completes."
 
-$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+if is_linux; then
+	set -x 
+	$DD if=/dev/urandom iflag=fullblock bs=1M count=$((64 * 1024 * 1024)) of="$TESTDIR/$TESTFILE1" &
+	ls -l $TESTDIR/$TESTFILE1
+else
+	$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+fi
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh]: Wasn't the plan to use `$FILE_WRITE` for this so we wouldn't need any linux specific logic?

File: tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh [hunk 11]
```diff
@@ -61,15 +62,21 @@ function cleanup
...
 log_assert "Turning both disks offline should fail."
 
 log_onexit cleanup
 
-$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+if is_linux; then
+        $DD if=/dev/urandom iflag=fullblock bs=1M count=$((64 * 1024 * 1024)) of=$TESTDIR/$TESTFILE1 &
+else
+        $FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+fi
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_002_neg.ksh]: Same, why not `$FILE_WRITE`?

File: tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh [hunk 12]
```diff
@@ -60,13 +61,21 @@ function cleanup
 
 	done
 
-	$KILL $killpid >/dev/null 2>&1
+	if $PS --no-headers -p $killpid > /dev/null; then
+		$KILL $killpid >/dev/null 2>&1
+	fi
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_001_pos.ksh]: I'm all for solid error handling but was adding this conditional really needed?  Even in the updated code we're always throwing away stderr, stdout, and the exit code.  Why not leave it as it was, the worst case for the existing case is it fails silently and harmlessly.  The worst case for the new code is somehow `ps` fails to list the pid and we don't even try to kill it, which is worse.  Or am I overlooking something?

File: tests/zfs-tests/tests/functional/online_offline/online_offline_003_neg.ksh [hunk 13]
```diff
@@ -67,15 +70,21 @@ create_pool $TESTPOOL1 $specials_list
 log_must $ZFS create $TESTPOOL1/$TESTFS1
 log_must $ZFS set mountpoint=$TESTDIR1 $TESTPOOL1/$TESTFS1
 
-$FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+if is_linux; then
+        $DD if=/dev/urandom iflag=fullblock bs=1M count=$((64 * 1024 * 1024)) of=$TESTDIR/$TESTFILE1 &
+else
+        $FILE_TRUNC -f $((64 * 1024 * 1024)) -b 8192 -c 0 -r $TESTDIR/$TESTFILE1 &
+fi
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/online_offline/online_offline_003_neg.ksh]: Same, why not `$FILE_WRITE?`.

File: tests/runfiles/linux.run [hunk 14]
```diff
@@ -98,13 +98,20 @@ tests = ['zfs_create_001_pos', 'zfs_create_002_pos', 'zfs_create_003_pos',
     'zfs_create_013_pos']
 
 # DISABLED:
+# zfs_destroy_001_pos - busy mountpoint behavior
 # zfs_destroy_005_neg - busy mountpoint behavior
 [tests/functional/cli_root/zfs_destroy]
+<<<<<<< 848a24ecacb5cb97541d53b6d5d55b947cce011d
```
Comment (rlaager) [inline tests/runfiles/linux.run]: There is a merge conflict left here.

Comment (behlendorf) [inline tests/runfiles/linux.run]: @jsalinasintel this merge conflict needs to be resolved.  Since the patch revolves around the `online_offline` there shouldn't need to be any changes here.

File: tests/zfs-tests/include/libtest.shlib [hunk 15]
```diff
@@ -37,7 +37,7 @@
 
 function is_linux
 {
-	if [[ $($UNAME -o) == "GNU/Linux" ]]; then
+    if [[ $($UNAME -o) == "GNU/Linux" ]]; then
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: You've got a gratuitous white space change here which should be dropped.

File: tests/zfs-tests/include/libtest.shlib [hunk 16]
```diff
@@ -2858,28 +2858,31 @@ function block_device_wait
 	fi
 }
 
-#
-# Synchronize all the data in pool
-#
-# $1 pool name
-#
-function sync_pool #pool
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: Why is sync_pool removed?  This is needed by several test cases.

File: tests/zfs-tests/include/libtest.shlib (hunk 15 above)
Comment (richardelling) [inline tests/zfs-tests/include/libtest.shlib]: The -o option to uname is not standardized across OSes, so you'll need to check that it fails outright.
How about this?
```
A=$($UNAME -o 2>/dev/null) || return 1
[[ $A == "GNU/Linux" ]] || return 1
return 0
```

File: tests/zfs-tests/include/libtest.shlib (hunk 16 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: Wow I have no idea.  I am going to close out this PR and start over I did not touch this part of the code at all

File: tests/zfs-tests/include/libtest.shlib (hunk 15 above)
Comment (jsalinasintel) [inline tests/zfs-tests/include/libtest.shlib]: I am actually not trying to change this here we decided to spit out into a different pull request
